packages/portalnetwork/src/wire/utp/PortalNetworkUtp/PortalNetworkUTP.ts

```diff
@@ -0,0 +1,495 @@
+          requestCode,
+          contentKeys,
+          sockets,
+          socketKey,
+          contents
+        )
+
+        if (this.openHistoryNetworkRequests[socketKey]) {
+          this.logger(`Request already Open`)
+        } else {
+          this.openHistoryNetworkRequests[socketKey] = newRequest
+          this.logger(`Opening request with key: ${socketKey}`)
+          await newRequest.init()
+        }
+
+        break
+      case 3:
+        sndId = connectionId + 1
+        rcvId = connectionId
+        socketKey = createSocketKey(peerId, sndId, rcvId)
+        if (this.openHistoryNetworkRequests[socketKey]) {
+          this.logger(`Request already Open`)
+        } else {
+          this.logger(`Opening request with key: ${socketKey}`)
+          sockets = contentKeys.map((k) => {
+            return this.createPortalNetworkUTPSocket(requestCode, peerId, sndId, rcvId)!
+          })
+          newRequest = new HistoryNetworkContentRequest(
+            requestCode,
+            contentKeys,
+            sockets,
+            socketKey,
+            []
+          )
+          this.openHistoryNetworkRequests[socketKey] = newRequest
+          await newRequest.init()
+        }
+        break
+    }
+    // const rcvId = requestCode % 2 === 0 ? sndId - 1 : sndId + 1
+    // const socket = this.createPortalNetworkUTPSocket(requestCode, peerId, sndId, rcvId, content)
+  }
+
+  createPortalNetworkUTPSocket(
+    requestCode: RequestCode,
+    peerId: string,
+    sndId: number,
+    rcvId: number,
+    content?: Uint8Array
+  ): UtpSocket | undefined {
+    let socket: UtpSocket
+    switch (requestCode) {
+      case 0:
+        socket = this.protocol.createNewSocket(
+          peerId,
+          sndId,
+          rcvId,
+          randUint16(),
+          0,
+          1,
+          undefined,
+          'write',
+          this.logger,
+          content
+        )
+        return socket
+      case 1:
+        socket = this.protocol.createNewSocket(
+          peerId,
+          sndId,
+          rcvId,
+          0,
+          randUint16(),
+          undefined,
+          1,
+          'read',
+          this.logger
+        )
+        return socket
+      case 2:
+        socket = this.protocol.createNewSocket(
+          peerId,
+          sndId,
+          rcvId,
+          0,
+          randUint16(),
+          undefined,
+          1,
+          'write',
+          this.logger,
+          content
+        )
+        return socket
+      case 3:
+        socket = this.protocol.createNewSocket(
+          peerId,
+          sndId,
+          rcvId,
+          randUint16(),
+          0,
+          1,
+          undefined,
+          'read',
+          this.logger
+        )
+        return socket
+    }
+  }
+
+  async handleUtpPacket(packetBuffer: Buffer, srcId: string): Promise<void> {
+    const requestKey = this.getRequestKeyFromPortalMessage(packetBuffer, srcId)
```

Comment:
What happens if we don't have a request key?  I know the `getRequestKeyFromPortalMessage` logs an error but should we do anything else (e.g. send an empty talkresp - or is that already happening)?